### DISCUSSION TITLE
Working with very large image - how do I max out all resource limits?

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#6068] [Help] [answered]
### ImageMagick version

7.0.8-Q16-HDRI

### Operating system, version and so on

Windows 10

### Description

Hi everyone. I have a task that involves working with a massive tiff file. I'm running into this error: "No space left on device @ error/cache.c/OpenPixelCache/3897. Does that mean disk space?

Some quick research is leading me in the direction of lifting the limits set in the policy.xml file. I noticed there are two policy.xml files. Not sure where I should be looking.

...\ImageMagick-7.0.8-Q16-HDRI\policy.xml
...\ImageMagick-7.0.8-Q16-HDRI\www\source\policy.xml

Is there a way to simply tell IM just max out everything my systems has in way of memory and diskspace? I am running ImageMagick on a PC so I am really not concerned with the policy in way of security and when Im running this task I dont need to conserve any resources because its a priority for me at that moment. Any help is appreciated here.

**Comment by snibgo** [comment]
>  I'm running into this error: "No space left on device @ error/cache.c/OpenPixelCache/3897. Does that mean disk space?

Yes, it probably means you have run out of space on your disk, and this is probably the disk used for %TEMP% on Windows or \tmp on unix. The remedy is to free space on that disk, or attach an external disk and tell IM to use that.

Try `magick -list resource`. If memory or disk have limits, you can edit your policy.xml to remove those limits.

**Comment by ghost** [comment]
Got it. Thank you!

**Comment by ghost** [comment]
I also read somewhere about deleting the policy.xml file to remove any limits. Curious to know anyone's thoughts about that approach.

**Comment by snibgo** [comment, marked as answer]
That works, but I wouldn't recommend it. Instead, I suggest editing policy.xml to comment-out each entry: `<!-- ... -->`.

**Comment by ghost** [comment]
Okay, thank you for this information.

**Comment by ghost** [comment]
I discovered that the policy.xml file is already commented out by default. That's good, but I wonder how IM assigns limits in this case because when I run magick -list resource there are limits defined. In anycase, I am now going to experiment with environment variable at command line for example, MAGICK_DISK_LIMIT to be sure limits are lifted.

**Comment by snibgo** [comment]
What limits are defined for you?

My policy.xml defines no limits. `magick -list resource` for me shows:
```
Resource limits:
  Width: 461.169PP
  Height: 461.169PP
  Area: 16.9485GP
  List length: unlimited
  Memory: 7.89224GiB
  Map: 15.7845GiB
  Disk: unlimited
  File: 1536
  Thread: 1
  Throttle: 0
  Time: unlimited
```
The only real limit there is on memory.

My 16 GB Windows laptop currently has 10 GB free, so IM won't use all of that. I don't know why, but I like the behaviour. I sometimes accidentally tell IM to do something daft that in the past made IM eat all the memory. When that happened my machine would hang, and no software could do anything, and maybe my machine would come back after 30 minutes or maybe I had to do a hard reboot with an unbent paperclip.

So, for me, IM not eating all the memory is a good thing.

**Comment by ghost** [comment]
Yes, I completely agree thats a good thing. But in this instance I am prioritizing this task in IM at all costs so I dont want any compromise in recourses - just want to give IM all it can take.

Here is what I have by default...

Resource limits:
  Width: 107.374MP
  Height: 107.374MP
  List length: -1
  Area: 8.4301GP
  Memory: 3.92557GiB
  Map: 7.85114GiB
  Disk: unlimited
  File: 1536
  Thread: 4
  Throttle: 0
  Time: unlimited

So the defaults are some decent limits and even though Im working with a big image looks like Im good for the most part. The issue I was facing was just the fact that I was running out of disk space but I am able to solve that by clearing up some space as you suggested earlier. With that said, I could make some changes to resources using environment variables for the better. I will increase my memory since I have 16Gib, and maybe increase the rest of the limits for good measure.

I noticed "-1" and "unlimited" used. Could I use either just to tell IM to use everything available or do I need to input exact amounts? Also can you tell me what "PP" stand for?

**Comment by snibgo** [comment]
When something has no limit, just ensure there isn't an entry for that thing.

P = pixel, B = byte.

"PP" is peta pixel. 1 peta = 1000 tera = 1000000 giga etc.

The "i" would mean the binary equivalent: 1 Pi = 1024 Ti = 1024*1024 Gi etc.
